mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/persistence/PersistableState.kt

```diff
@@ -0,0 +1,246 @@
+
+import jetbrains.mps.project.AbstractModule
+import kotlinx.coroutines.runBlocking
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.buildClassSerialDescriptor
+import kotlinx.serialization.descriptors.element
+import kotlinx.serialization.encoding.CompositeDecoder.Companion.DECODE_DONE
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.decodeStructure
+import kotlinx.serialization.encoding.encodeStructure
+import mu.KotlinLogging
+import org.jetbrains.mps.openapi.persistence.PersistenceFacade
+import org.modelix.kotlin.utils.UnstableModelixFeature
+import org.modelix.model.client2.ModelClientV2
+import org.modelix.model.lazy.BranchReference
+import org.modelix.model.lazy.CLVersion
+import org.modelix.model.lazy.RepositoryId
+import org.modelix.mps.sync.IBinding
+import org.modelix.mps.sync.IRebindModulesSyncService
+import org.modelix.mps.sync.bindings.BindingsRegistry
+import org.modelix.mps.sync.modelix.BranchRegistry
+import org.modelix.mps.sync.mps.ActiveMpsProjectInjector
+import org.modelix.mps.sync.mps.notifications.InjectableNotifierWrapper
+import org.modelix.mps.sync.transformation.cache.MpsToModelixMap
+
+@UnstableModelixFeature(reason = "The new modelix MPS plugin is under construction", intendedFinalization = "2024.1")
+/**
+ * States are capable of taking a snapshot of the current bindings to modelix servers with getCurrentState(),
+ * and recreating that state by calling load(). Note that currently the bindings of the state will be added during
+ * load(), but not included bindings will be disconnected.
+ *
+ * States will be automatically saved when a project is closed, and automatically loaded, when that project is
+ * reopened. Technically, it should also be possible to create a state yourself in order to load it, although this
+ * is not what they were made for.
+ *
+ * WARNING:
+ * About the State class:
+ * - do not make it an inner class, otherwise deserialization will fail with an exception
+ * About the State's fields:
+ * - Mutable collections will not be persisted!!!
+ * - Maps and Collections will only be persisted 2 layers deep (List<List<String>> works, but List<List<List<String>>> not)
+ * - Pairs will not be persisted
+ *
+ * @property clientUrl the URL of the model server to which we were connected
+ * @property repositoryId the ID of the modelix repository to which we were connected
+ * @property branchName the name of the branch in the modelix branch
+ * @property localVersion the latest version hash we know from the branch
+ * @property moduleIds the IDs of the Modules that were synchronized
+ * @property synchronizationCache the [MpsToModelixMap] serialized into a string
```

Comment:
* `synchronizationCache` does not exist anymore.

Reply:
fixed in 87ce51b6122bd0e0e5a77c2a9a0970031eb02db1